script/tool/lib/src/common/plugin_command.dart

```diff
@@ -266,15 +275,50 @@ abstract class PluginCommand extends Command<void> {
   ///    is a sibling of the packages directory. This is used for a small number
   ///    of packages in the flutter/packages repository.
   Stream<PackageEnumerationEntry> _getAllPackages() async* {
+    final Set<String> packageSelectionFlags = <String>{
+      _packagesArg,
+      _runOnChangedPackagesArg,
+      _packagesForBranchArg,
+    };
+    if (packageSelectionFlags
+            .where((String flag) => argResults!.wasParsed(flag))
+            .length >
+        1) {
+      printError('Only one of --$_packagesArg, --$_runOnChangedPackagesArg, or '
+          '--$_packagesForBranchArg can be provided.');
+      throw ToolExit(exitInvalidArguments);
+    }
+
     Set<String> plugins = Set<String>.from(getStringListArg(_packagesArg));
 
+    final bool runOnChangedPackages;
+    if (getBoolArg(_runOnChangedPackagesArg)) {
+      runOnChangedPackages = true;
+    } else if (getBoolArg(_packagesForBranchArg)) {
+      final String? branch = await _getBranch();
+      if (branch == null) {
+        printError('Unabled to determine branch; --$_packagesForBranchArg can '
+            'only be used in a git repository.');
+        throw ToolExit(exitInvalidArguments);
+      } else {
+        runOnChangedPackages = branch != 'master';
+        // Log the mode for auditing what was intended to run.
+        print('--$_packagesForBranchArg: running on '
+            '${runOnChangedPackages ? 'changed' : 'all'} packages');
+      }
+    } else {
+      runOnChangedPackages = false;
+    }
+
     final Set<String> excludedPluginNames = getExcludedPackageNames();
 
-    final bool runOnChangedPackages = getBoolArg(_runOnChangedPackagesArg);
-    if (plugins.isEmpty &&
-        runOnChangedPackages &&
-        !(await _changesRequireFullTest())) {
-      plugins = await _getChangedPackages();
+    if (runOnChangedPackages) {
+      final GitVersionFinder gitVersionFinder = await retrieveVersionFinder();
+      final List<String> changedFiles =
+          await gitVersionFinder.getChangedFiles();
+      if (!_changesRequireFullTest(changedFiles)) {
```

Comment:
Sure, I can change it in a follow-up. Neither the API nor the way it's being used are new in this PR, I'm just rearranging the flow a bit.

Reply:
Actually, having looked at the details I'm not clear how it would be more efficient. `getChangedFiles` is taking the full stdout from `git` (the process API for the tools aren't stream-based) and splitting it by lines, giving a list. So returning a `Stream` instead of a `List` doesn't change anything there.

At this call site I need this list in two places: `_changesRequireFullTest` and `_getChangedPackages`. I can't consume the stream twice, so I would need to `.toList()` it here.

So the difference would be that instead of taking a `List` and returning it to a caller that needs a `List`, I would be taking a `List`, converting it to a `Stream`, and giving the stream to the caller to immediately turn back to a `List`. Am I missing a different way of structuring this?

Reply:
Yea, if you are going to cache the result it doesn't make a difference, it's more of an academic change.

Returning a Stream is better because it allows people to process the output while it is being parsed.  This is especially helpful when you are working on input from files/pipes.  This also means the operation isn't memory bound and if a consumer decides to stop consuming mid-stream  resources won't be wasted processing the output that happens after it stops.

It doesn't mean a lick of difference if you cache the results though so you can process it twice.

Reply:
> Returning a Stream is better because it allows people to process the output while it is being parsed. This is especially helpful when you are working on input from files/pipes.

The supply-side issue here is that the `git` package doesn't steam `git` output, it uses `run` and returns the full output as a String.